### PR TITLE
Pull request for merging Modulus function

### PR DESCRIPTION
**Pull request by Simi-Solola**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/modulus.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    b (float): The denominator (non-zero).
+
+    Returns:
+    float: The remainder of a divided by b.
+
+    Raises:
+    ValueError: If the denominator is zero.
+    """
+    if b == 0:
+        raise ValueError("The denominator (b) cannot be zero.")
```
Comment (Ridwanayinde) [inline solutions/modulus.py:31]: Kudos, @Simi-Solola. You can improve the statement in the error you raised to make it clearer, like " The numerator and denominator can be any real numbers except the denominator cannot be zero."

File: solutions/tests/test_modulus.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#! usr/bin/env python3
```
Comment (Ridwanayinde) [inline solutions/tests/test_modulus.py:1]: Here, you have a typo error here. It should have been #!/usr/bin/env python3

File: solutions/tests/test_modulus.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        """Test modulus when the denominator is zero."""
+        with self.assertRaises(ValueError):
+            modulus(10, 0)
+
+    def test_float_inputs(self):
+        """Test modulus with floating-point numbers."""
+        self.assertAlmostEqual(modulus(10.5, 3), 1.5)
+        self.assertAlmostEqual(modulus(-10.5, 3), -1.5)
+        self.assertAlmostEqual(modulus(-10.5, -3), -1.5)
+        self.assertAlmostEqual(modulus(10.5, -3), 1.5)
```
Comment (Ridwanayinde) [inline solutions/tests/test_modulus.py:49]: Add boundary test for very large and very small numbers e.g (modulus (1e18,3), 1), where the numerator is very large. and the test the denominator where the denominator is very close to zero but still valid e.g. (modulus(1, 1e-10), 0). This is to ensure the function behaves as expected in cases where rare bugs could occur when it is tested against very small or large number in real scenario.


Also, you can add one or two defensive tests apart from when the denominator is zero. This is just to make your test to be more robust. Like when the numerator is a string or when the user inputs **None** as the denominator.

**Review by Ridwanayinde — COMMENTED** [review]
Hi @Simi-Solola.  I am taking on the review of your code.

File: solutions/tests/test_modulus.py (hunk 3 above)
Comment (Simi-Solola) [reply solutions/tests/test_modulus.py:49]: This sounds very reasonable. thanks a lot Ridwan

File: solutions/modulus.py (hunk 1 above)
Comment (Simi-Solola) [reply solutions/modulus.py:31]: Alright, I’d do as advised

File: solutions/tests/test_modulus.py (hunk 2 above)
Comment (Simi-Solola) [reply solutions/tests/test_modulus.py:1]: Love your attention to details! Thank you

File: solutions/modulus.py (hunk 1 above)
Comment (Simi-Solola) [reply solutions/modulus.py]: @Ridwanayinde I have corrected this. Kindly check it out. Will you also be the one to 'resolve the conversation'?

File: solutions/tests/test_modulus.py (hunk 2 above)
Comment (Simi-Solola) [reply solutions/tests/test_modulus.py:1]: @Ridwanayinde I have corrected this as well

File: solutions/tests/test_modulus.py (hunk 3 above)
Comment (Simi-Solola) [reply solutions/tests/test_modulus.py:50]: Done. KIndly check @Ridwanayinde

File: solutions/tests/test_modulus.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#! usr/bin/env python3
+#!usr/bin/env python3
```
Comment (Ridwanayinde) [inline solutions/tests/test_modulus.py:1]: I'm sorry, @Simi-Solola. You have tried, but this code is failing the py formatting. It could be shebang #!/usr/bin/env python3. You missed the forward slash. Kindly check it out.

**Review by Ridwanayinde — CHANGES_REQUESTED** [review]
@Simi-Solola. The code is okay now but kindly make sure the code passes the CI checks. You can make use of the ruff formating to format the code.

**Review by Ridwanayinde — DISMISSED** [review]

**Comment by Ridwanayinde** [comment]
@Simi-Solola, I am about to merge now, but I realize you did not include any doctest in your docstring. Please, kindly do it.

**Comment by Simi-Solola** [comment]
@Ridwanayinde I have done that now please 🥹

**Review by Ridwanayinde — APPROVED** [review]
Great work @Simi-Solola
